plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/HostedMachineProviderImpl.java

```diff
@@ -215,4 +267,22 @@ protected void buildImage(CheServiceImpl service,
             }
         }
     }
+
+    @PreDestroy
+    private void preDestroy() {
+        finalizeSnapshotImagesCleaner();
+    }
+
+    /**
+     * Tries to clean up snapshots images which is in the cleaner queue, but were canceled due to server stopping.
+     * To avoid freezing of server on stop performs post clean in separate thread.
+     */
+    private void finalizeSnapshotImagesCleaner() {
+        List<Runnable> queue = snapshotImagesCleanerService.shutdownNow();
+        Thread snapshotsCleaner = new Thread(() -> queue.forEach(Runnable::run));
+        snapshotsCleaner.setDaemon(true);
+        snapshotsCleaner.setName("PostSnapshotImagesCleaner");
```

Comment:
Consider another name, for example PreDestroySnapshotImagesCleaner, FinalizingSnapshotImagesCleaner, SnapshotImagesCleaningFinalizer